### PR TITLE
Story/asa fuel layer data/2208

### PR DESCRIPTION
**Pull request by andrea-williams**
Data for all fire zones sent from API to FE (visible in Network tab). Subsequent ticket(s) will display the data in the UI, and allow for fire zone selection so that query can be sped up by only fetching data for relevant fire zone. (Alternatively, could integrate fuel type analysis into SFMS processing and store results in DB rather than doing the calculations on the fly.)

Also replaced `run_date` (which was date-only) with `run_datetime` (datetime) so that we can distinguish between AM & PM SFMS runs on forecast data.
# Test Links:
[Percentile Calculator](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-2554.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: api/app/routers/fba.py [hunk 1]
```diff
@@ -33,31 +34,64 @@ async def get_all_fire_centers(_=Depends(authentication_required)):
...
+    async with get_async_read_session_scope() as session:
+        fuel_types_high_hfi = await get_fuel_types_with_high_hfi(
+            session,
+            run_type=RunTypeEnum(run_type.value),
+            for_date=for_date,
+            run_datetime=run_datetime
+        )
+        fire_zones_hfi_fuel_types = []
+        current_zone_id = None
+        for row in fuel_types_high_hfi:
```
Comment (conbrad) [inline api/app/routers/fba.py:77]: What is this loop doing?

Comment (andrea-williams) [reply api/app/routers/fba.py:77]: iterating through the list of raw SQL row results and building out a response in the format of `List[FireZoneHfiThresholdsByFuelTypes]`

Comment (andrea-williams) [reply api/app/routers/fba.py:77]: example from my local:
```
{
  "mof_fire_zone_id": 401,
  "fuel_types": [
    {
      "fuel_type_id": 2,
      "threshold": 2,
      "area": 413180.88182108715
    },
    {
      "fuel_type_id": 3,
      "threshold": 1,
      "area": 16036589.21019188
    },
    {
      "fuel_type_id": 3,
      "threshold": 2,
      "area": 0
    },
    {
      "fuel_type_id": 4,
      "threshold": 1,
      "area": 13854687.478280807
    },
    {
      "fuel_type_id": 5,
      "threshold": 1,
      "area": 0
    },
    {
      "fuel_type_id": 5,
      "threshold": 2,
      "area": 0
    },
    {
      "fuel_type_id": 6,
      "threshold": 1,
      "area": 0.000003965822036318498
    },
    {
      "fuel_type_id": 7,
      "threshold": 1,
      "area": 64173011.49065131
    },
    {
      "fuel_type_id": 7,
      "threshold": 2,
      "area": 172.81822308288335
    },
    {
      "fuel_type_id": 8,
      "threshold": 1,
      "area": 8.85075916702494
    },
    {
      "fuel_type_id": 8,
      "threshold": 2,
      "area": 0
    },
    {
      "fuel_type_id": 9,
      "threshold": 1,
      "area": 0
    },
    {
      "fuel_type_id": 12,
      "threshold": 1,
      "area": 173.70439270479034
    },
    {
      "fuel_type_id": 12,
      "threshold": 2,
      "area": 171.4817892202029
    },
    {
      "fuel_type_id": 14,
      "threshold": 1,
      "area": 264783760.6426373
    },
    {
      "fuel_type_id": 14,
      "threshold": 2,
      "area": 80245661.31049424
    },
    {
      "fuel_type_id": 99,
      "threshold": 1,
      "area": 1.212138028838794e-16
    },
    {
      "fuel_type_id": 99,
      "threshold": 2,
      "area": 0
    },
    {
      "fuel_type_id": 102,
      "threshold": 1,
      "area": 0
    }
  ]
}
```

Comment (conbrad) [reply api/app/routers/fba.py:77]: Makes me nervous how many conditionals and state mutations are in there, why not just:
```
        fuel_stats_by_fire_zone = groupby(fuel_types_high_hfi, operator.itemgetter(0))
        fire_zone_stats = dict((k, list(map(lambda x: x, values)))
                               for k, values in fuel_stats_by_fire_zone)
```
Then iterate over the existing dictionary of `fire_zone_id` ->  (`fuel_type_id`, `threshold`, `area`)

Comment (andrea-williams) [reply api/app/routers/fba.py:77]: because I prefer to live and code dangerously! 😈

jk, I'll make that change now

Comment (conbrad) [reply api/app/routers/fba.py:77]: Here it is: https://github.com/bcgov/wps/compare/story/asa-fuel-layer-data/2208...task/asa-fuel-layer-data-endpoint?expand=1

Comment (andrea-williams) [reply api/app/routers/fba.py]: @conbrad  what's that link for? When I click it, I'm getting a prompt to open a new PR

Comment (andrea-williams) [reply api/app/routers/fba.py]: FYI I just pushed a change to implement your feedback

Comment (conbrad) [reply api/app/routers/fba.py]: > @conbrad what's that link for? When I click it, I'm getting a prompt to open a new PR

Scroll down to see the diff that implements a fully working version of what I was talking about

Comment (conbrad) [reply api/app/routers/fba.py]: But yeah looks like you got it, nice!

File: api/app/routers/fba.py [hunk 2]
```diff
@@ -67,13 +98,14 @@ async def get_run_datetimes_for_date_and_runtype(run_type: RunType, for_date: da
...
         for row in rows:
-            datetimes.append(row.run_datetime)
+            datetimes.append(row.run_datetime)  # type: ignore
 
         return datetimes
 
 
 @router.get('/fire-zone-hfi-areas/{run_type}/{run_date}/{for_date}',
             response_model=FireZoneHighHfiAreasListResponse)
+# TODO: this function doesn't appear to be being used, and is calling itself...
```
Comment (dgboss) [inline api/app/routers/fba.py:108]: Yep, it's an accidental artifact and can be deleted.

**Review by conbrad — APPROVED** [review]

**Review by dgboss — APPROVED** [review]

**Pull request merged**